tests/test_modeling_perceiver.py

```diff
@@ -196,6 +193,24 @@ def get_config(self):
             num_labels=self.num_labels,
         )
 
+    def update_config_with_model_class(self, config, model_class):
+        # Perceiver is a bit specific since `d_model` needs to be defined
+        # ahead of time in the config, but the actual desired `d_model` might
+        # depend on the model class we want to use so this is an escape hatch
+        # to enable adjusting `d_model` after we know which head we intend to
+        # use
+        if model_class.__name__ == "PerceiverForImageClassificationLearned":
+            config.d_model = 512
+        elif model_class.__name__ == "PerceiverForImageClassificationFourier":
+            config.d_model = 261
+        elif model_class.__name__ == "PerceiverForImageClassificationConvProcessing":
+            config.d_model = 322
+        elif model_class.__name__ == "PerceiverForOpticalFlow":
+            config.d_model = 322
+        elif model_class.__name__ == "PerceiverForMultimodalAutoencoding":
+            config.d_model = 409
+        return config
```

Comment:
Not a huge fan of this but I understand why it has to be done

Reply:
I don't like it either, but IMO it was even worse to make `get_config` or `get_pipeline_config` depend on `model_class`.

These values are actually linked to the default values of `PerceiverFeatureExtractor` (1, 3, 224, 224) and the other defaults of the config:

https://github.com/huggingface/transformers/blob/master/src/transformers/models/perceiver/modeling_perceiver.py#L2889 

This is the calculation that could be done to infer `d_model` from the other arguments. But this depends on `prep_type` which is an attribute of PreProcessor (not of the config).

Reply:
Resolved with the fix https://github.com/huggingface/transformers/pull/14750